3-enrich/beam-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/utils.scala

```diff
@@ -0,0 +1,167 @@
+
+import scala.util.Try
+
+import org.joda.time.{DateTime, DateTimeZone}
+import org.joda.time.format.DateTimeFormat
+import org.json4s.{JObject, JValue}
+import org.json4s.JsonDSL._
+import org.json4s.jackson.Serialization.write
+import scalaz._
+
+import common.enrichments.EnrichmentRegistry
+import common.outputs.{EnrichedEvent, BadRow}
+import iglu.client.validation.ProcessingMessageMethods._
+import singleton._
+
+object utils {
+
+  /** Format an [[EnrichedEvent]] as a TSV. */
+  def tabSeparatedEnrichedEvent(enrichedEvent: EnrichedEvent): String =
+    enrichedEvent.getClass.getDeclaredFields
+    .filterNot(_.getName.equals("pii"))
+    .map { field =>
+      field.setAccessible(true)
+      Option(field.get(enrichedEvent)).getOrElse("")
+    }.mkString("\t")
+
+  private val formatter = DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSS")
+  private implicit val formats = org.json4s.DefaultFormats
+  /** Creates a PII event from the pii field of an existing event. */
+  def getPiiEvent(event: EnrichedEvent): Option[EnrichedEvent] =
+    Option(event.pii)
+      .filter(_.nonEmpty)
+      .map { piiStr =>
+        val ee = new EnrichedEvent
+        ee.unstruct_event = event.pii
+        ee.app_id = event.app_id
+        ee.platform = "srv"
+        ee.etl_tstamp = event.etl_tstamp
+        ee.collector_tstamp = event.collector_tstamp
+        ee.event = "pii_transformation"
+        ee.event_id = UUID.randomUUID().toString
+        ee.derived_tstamp = formatter.print(DateTime.now(DateTimeZone.UTC))
+        ee.true_tstamp = ee.derived_tstamp
+        ee.event_vendor = "com.snowplowanalytics.snowplow"
+        ee.event_format = "jsonschema"
+        ee.event_name = "pii_transformation"
+        ee.event_version = "1-0-0"
+        ee.v_etl = s"beam-enrich-${generated.BuildInfo.version}-common-${generated.BuildInfo.sceVersion}"
+        ee.contexts = write(getContextParentEvent(ee.event_id))
+        ee
+      }
+
+  private def getContextParentEvent(eventId: String): JValue =
+    ("schema" -> "iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0") ~
+      ("data" -> List(
+        ("schema" -> "iglu:com.snowplowanalytics.snowplow/parent_event/jsonschema/1-0-0") ~
+          ("data" -> ("parentEventId" -> eventId))))
+
+  def emitPii(registry: EnrichmentRegistry): Boolean =
+    registry.getPiiPseudonymizerEnrichment.map(_.emitIdentificationEvent).getOrElse(false)
+
+  /**
+   * Truncate an oversized formatted enriched event into a [[BadRow]].
+   * @param enrichedEvent TSV-formatted oversized enriched event
+   * @param bytesSize size in bytes of the formatted enriched event
+   * @param maxBytesSize maximum size in bytes a record can take
+   * @return a [[BadRow]] containing a the truncated enriched event (10 times less than the max size)
+   */
+  def resizeEnrichedEvent(enrichedEvent: String, bytesSize: Int, maxBytesSize: Int): BadRow =
+    BadRow(enrichedEvent.take(maxBytesSize / (4 * 10)), NonEmptyList(
```

Comment:
I don't understand this number (/(4*10)) in this context (I am thinking of 39 max bytes for instance). Could we make it a constant with a descriptive name? Also, would it not be better to guarantee some fields make it to the truncated BadRow (for instance eventId)?

Reply:
good point, I added a comment to the constant saying: "We want to take one-tenth of the payload characters and one character can take up to 4 bytes"